Basic Bundles Data Models

ormsbee opened the pull request:
Basic Bundles, management commands to create/update Bundles, and Django Admin functionality to inspect them.

ormsbee commented on blockstore/apps/bundles/data/store.py:
```diff
@@ -0,0 +1,222 @@
+    public BundleStore interface. To keep the option open, we'd want to make
+    sure that things going in and out of BundleStore don't leak storage
+    implementation details.
+    """
```
I've been mucking around with version modeling all day, and I think I'm landing on a different separation than what was originally put in the proposal. It basically involves pushing version number understanding up to the models.py level, while keeping the stuff below that in the BundleDataStore purely content-hash based (the BundleVersion model bridges the two).

Haven't quite figured out the implications on dependency calculation. It might have some positive implications for content removal (i.e. the separation allows us to remove content snapshots without necessarily removing the BundleVersion rows).

Anyhow, a lot to munch on, and it's probably pushed this work out a day. But I think the separation will be good for both separation of concerns and operational concurrency issues.

symbolist replied:
Have you considered including the content address of the previous snapshot in the summary to create a chain and maintaining ordering at the snapshot level? If we did that then the version "numbers" would become derived data (though we may still want to keep them in the model for efficient reads).

ormsbee replied:
It's crossed my mind, but I'm wary of adding more version control features. Right now, a new snapshot can be created without knowing anything about the previous snapshot, and a split where two entries have the same parent manifests in two versions being created. It's not smart, but it's simple.

I'm definitely open to it down the line, especially if it's needed to make drafts work well. But I'd rather hold off until we have a concrete use for tracking that level.

ormsbee commented on blockstore/apps/bundles/data/store.py:
```diff
@@ -0,0 +1,222 @@
+    both see the database reflecting the same state. But if BundleStore creates
+    only hash-versioned Snapshots, then it can do all that work and the only
+    time version_num is read and incremented for a new entry is when
+    BundleDataStore emits a signal that BundleVersion catches.
```
Actually, I guess since the only thing that needs version information in the BundleStore is the summary JSON, we still wouldn't have much in the way of concurrency issues because we could delay until then. But that would involve it calling up the abstraction stack a layer, which still feels wrong.

symbolist commented on blockstore/apps/bundles/data/store.py:
```diff
@@ -0,0 +1,222 @@
+import attr
+
+
+snapshot_created = Signal(providing_args=["bundle_uuid", "snapshot_hash"])
```
Since we are going for immutable data structures (👍), I would recommend that instead of (or at least in addition to) signals, a transaction log is maintained (it could be a simple Django model). This would allow plugins to operate out of process and asynchronously and have far more robust error recovery than possible via patterns like this: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/tasks.py#L95-L122.

ormsbee replied:
Yeah, transaction log sounds great. I didn't take a stab at it here because there are a few separate-but-related use cases and it's not clear to me what the data model should look like:

**Security audit log.** Permanent storage, search by user or resource. Possibly by time range. Covers changes to mutable data structures (e.g. Collections) as well as Bundle creation. Storage in Django models.

**Who is working on my project now?** Long-asked for Studio feature to show who was recently or who is actively working on a course so they don't step on each others toes. Long term storage probably not necessary, queries more likely at collection level over a limited time range. Might be serviceable using the security audit log model.

**Replay-style transaction logs for plugins.** Verbose, query either the most recent thing or possibly a large chunk. This sounds very much like a fit for Kafka, but Django models and a read-replica might work.

So I agree in principle that we should add this eventually, but I'm not sure exactly what form it should take. I hope that gets clearer as we develop more plugins.

bradenmacdonald commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,186 @@
+    """
+    id = models.BigAutoField(primary_key=True)
+
+    # Would love to have better names here
```
`lender` and `borrower` are probably the least ambiguous though, which is a very nice property when it comes to links. Maybe `source` and `consumer` if you just want some suggestions on different words though.

bradenmacdonald commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,186 @@
+    # We use a BinaryField(16) because Django's built-in UUID field is double
+    # the size it needs to be (32 vs. 16 bytes) on MySQL. We use url-safe base64
+    # encoding when passing around serialized references.
+    uuid = models.UUIDField(default=uuid.uuid4, unique=True)
```
I imagine you were already thinking of these but:

This says we use `BinaryField(16)` but it looks to be using the standard char storage?

Should we have a mixin that defines these `id BigAutoField, uuid UUIDField` fields and move the explanations about MySQL's limitations to that file?

bradenmacdonald commented on blockstore/apps/bundles/migrations/0004_bundleversion_snapshot_hash.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.14 on 2018-07-26 01:14
```
Note: let's collapse these migrations before we finish this PR.

ormsbee commented:
Okay, I moved around a lot of stuff, but I think the basic structure and logic are in a reviewable state. I have simple commands to create Bundles and add BundleVersions, and tests around it. Also made a few yak-shaving updates where the cookiecutter hasn't caught up to more recent Open edX best practices (removed nose and bower, added pytest).

If you've run this code before, please wipe your DB (or just your bundles tables). I deleted the migrations that were evolving over the various commits and regenerated a single migration.

symbolist commented on blockstore/apps/bundles/store.py:
```diff
@@ -0,0 +1,232 @@
+                'bundle_uuid': obj.bundle_uuid,
+                'hash_digest': obj.hash_digest.hex(),
+                'files': obj.files,
+            }
```
A document version would be useful in case we need to make a backwards-incompatible change.

ormsbee replied:
Oops. Yeah, and that was even in the design doc... Will add, thank you!

symbolist commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,223 @@
+
+    BundleVersion ID serialization could be left to the view layer, since we're
+    not storing serialized BundleVersion IDs anywhere as one thing. It's mostly
+    just input shorthand.
```
Hmm. In general I prefer to defer additive things unless there is a purpose for them. It may be needed depending on how links to other bundles are specified.

ormsbee replied:
I'll take it out for now. It's not that hard to add it back in later if we start running into it.

ormsbee commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,223 @@
+    #     Collection, related_name="bundles", related_query_name="bundle"
+    # )
+
+    slug = models.SlugField(allow_unicode=True)  # For pretty URLs
```
Yeah, I was thinking it would be something like api/bundles/<uuid>-<slug>, and that the slug would also be reflected in an export scenario so that you can see some readable stuff in a sea of UUID-named directories.

pomegranited replied:
+1 to that -- Slugs are just for humans, so it shouldn't matter if they change.

ormsbee commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,223 @@
+        """
+        The BundleStore layer doesn't know we exist, but it emits signals when a
+        new BundleSnapshot is created. We listen for that signal and create a
+        new BundleVersion that points to it.
```
> Is the Blockstore supposed to be sufficient as the primary storage for authoring interfaces?

Yes.

> 1. If yes, should every save-preview cycle result in a new version? Or if the author uploads a few files one by one to a bundle in an authoring session, should each file addition result in a new version?

Absolutely not. We've walked down that path with SplitMongo, and I have no desire to go down it again. I finally wrote a [script](https://github.com/edx/tubular/blob/master/tubular/scripts/structures.py) to clean up most of that garbage a year ago. There's no value in storing all of those as separate versions, and so we shouldn't.

I think there should be a draft directory where you read and write to your hearts content and nothing is committed to a snapshot or version. At some point, you decide, "yes, this is a new version right here", and then you press a button and it becomes a version. I'm not sure the exact form that will take in implementation -- do we copy everything from the most recent version when starting a draft? Lazy copy-on-write? What happens if a person creates a new version while we're still editing our draft? But yeah, definitely not thousands of useless versions. Ugh.

> 2. What purpose do version numbers serve in an authoring interface?

It's like the basic versioning information you can add in wikis (but nobody does in practice). They can associate messages describing the update, and they'd get a list of versions, dates, who did the editing, and commit messages (I realize most of those aren't there yet).

> Is the author going to get a long list of numbers or are there going to be any other annotations like tags?

The expectation would be that other things annotate BundleVersions, and that would include search and tagging.

> If the tags are going to be stored in the tagging service for example, will it possible for authors to use them for linking from one bundle to another or will they have to find out the corresponding version number and use that? Or should version tags be first class entities in the Blockstore?

Sorry, there are a few different types of tagging that have been floating around in different conversations and I'm not sure what you mean by that.

> 3. Can we add a created_at field? I think that is going to be very helpful.

Yeah. I punted on it for a moment because we weren't using it in a view yet and I was punting on whether to add the drf-extensions dependency or just use straight timestamps, but either way they should be in there. I'll add them.

> Is there any document which has a list of authoring workflows from the product perspective so we can see if/how the APIs will work in such scenarios?

@bradenmacdonald would likely have the latest and greatest on that front.

bradenmacdonald replied:
> Is there any document which has a list of authoring workflows from the product perspective so we can see if/how the APIs will work in such scenarios?

> @bradenmacdonald would likely have the latest and greatest on that front.

There's a WIP version out there but I haven't got access to it yet. Will share with y'all as soon as it's ready. The LX team has been making some major revisions.

symbolist replied:
> Absolutely not. We've walked down that path with SplitMongo, and I have no desire to go down it again. I finally wrote a script to clean up most of that garbage a year ago. There's no value in storing all of those as separate versions, and so we shouldn't.

Since you had added extensive comments about a lot of other things in the code, but nothing here, I was a bit unsure. * phew * ;) This sounds good. 

I have created a ticket so we can tackle this separately: https://tasks.opencraft.com/browse/LX-50

> It's like the basic versioning information you can add in wikis (but nobody does in practice). They can associate messages describing the update, and they'd get a list of versions, dates, who did the editing, and commit messages (I realize most of those aren't there yet).
> Sorry, there are a few different types of tagging that have been floating around in different conversations and I'm not sure what you mean by that.

I meant something like git tags that are used for referring to specific versions and can have some meaning. What I am getting is that if someone wants to add semantic versions  e.g: release-2018-01-01 or v2.3.3, they should put those in the description field and links between bundles must use the integer `version_num` for specifying the `BundleVersion`. Is that correct?

pomegranited commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,223 @@
+
+class Collection(models.Model):
+    """
+    Administrative grouping for Bundles: policy, permissions, licensing, etc.
```
I've shared a folder with you all containing some discussions done among a representative of Creative Commons, edX's OSCM at the time, a community member, and OpenCraft, which might help clarify some of the use cases requested for CC licensing.

(One of the topics that came up during those discussions was the desire for exporting/importing partial courses, which, of course, Blockstore achieves beautifully 😄 )

Having the ability to say "all versions of the Bundles in this Collection are now available under Creative Commons" gets tricky, if the license becomes more restrictive (e.g. adds a non-commercial clause), and existing links to the more open content bundles already exist.  So I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

>  It would also allow us to handle cases where things are licensed in multiple ways

Which "things"?  Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions).  And so the license of the Collection itself is an amalgam of the licenses of its Bundles.  Whether or not the Collection has its own license -- don't know if is necessary or not.

ormsbee replied:
> Which "things"? Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions). And so the license of the Collection itself is an amalgam of the licenses of its Bundles. Whether or not the Collection has its own license -- don't know if is necessary or not.

Sorry, that was really unclear of me. I was thinking that if a Collection (and by extension a Bundle) were available under multiple licenses simultaneously (e.g. GPL/BSD), we'd want some way to identify which license was being used on any particular borrowing. I agree that all the contents of a Bundle should share the same license. I would also like to take the tack that the entire Collection shares the same license.

> Having the ability to say "all versions of the Bundles in this Collection are now available under Creative Commons" gets tricky, if the license becomes more restrictive (e.g. adds a non-commercial clause), and existing links to the more open content bundles already exist. So I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model). In the case where licensing gets more restrictive, we'd still be able to look up all Bundles that borrowed under the more permissive license and do something about it if necessary. Otherwise, they would be allowed to use the old thing but wouldn't be allowed to upgrade.

This gives Links and Licensing a tighter relationship than in the original design, but I think that's okay, and makes sense if we're going to treat licensing seriously.

bradenmacdonald replied:
> I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model).

That sounds good to me.

> In the case where licensing gets more restrictive, we'd still be able to look up all Bundles that borrowed under the more permissive license and do something about it if necessary. Otherwise, they would be allowed to use the old thing but wouldn't be allowed to upgrade.

That would require that Blockstore "understand" the various licenses and their restrictions, right?

symbolist replied:
> I've shared a folder with you all containing some discussions done among a representative of Creative Commons, edX's OSCM at the time, a community member, and OpenCraft, which might help clarify some of the use cases requested for CC licensing.

Thanks! That is great. Will give it a read.

>  Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions).

A bundle may have text, javascript libs, images and videos taken from other places on the internet with different licenses. I imagine that would be very common. What is the license of the bundle in that case?

> I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model).

This breaks the common convention that the license of a version of an item is what was declared at the time it was released. If I am linking to bundle1@1 from bundle2 I should be able link to it from bundle3 even if bundle1@2 has a more restrictive license.

> I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

This makes more sense to me.

I wasn't part of the early discussions but my feeling is that the goal of tracking license info as a first class thing is to improve discovery of items for re-use (and to enforce restrictions) e.g. "this bundle can be used without modification", "this bundle requires attribution for re-use". I think it would be better if we first came up with a list of scenarios/use-cases that would need to be handled and then figure out what data model would work best. Makes sense?

ormsbee replied:
> That would require that Blockstore "understand" the various licenses and their restrictions, right?

Yeah. I'm still hoping the layer that knows is relatively thin and isolated, but yeah.

> A bundle may have text, javascript libs, images and videos taken from other places on the internet with different licenses. I imagine that would be very common. What is the license of the bundle in that case?

I don't think it's worth the cost to try to track every JavaScript library file in a Bundle and map out their licenses. Blockstore can support a handful of pre-baked licenses at the Collection level -- probably "all rights reserved, only my university gets to use this", and the various flavors of Creative Commons. If a user is going to put a separately licensed third-party piece of content into their Bundle, it should be their responsibility to make sure the licensing is compatible. It should also be their responsibility to include any information needed for proper attribution of authorship.

> This breaks the common convention that the license of a version of an item is what was declared at the time it was released. If I am linking to bundle1@1 from bundle2 I should be able link to it from bundle3 even if bundle1@2 has a more restrictive license.

As a developer mapping this onto the source code analogy, I agree. But I don't think that necessarily maps with product expectations. I looked through the docs @pomegranited shared, and I can't see any place this was hashed out.

My thinking was basically:

1. Going from a more permissive to more restrictive license is likely going to be an edge case.
2. If it's really a massive shift-over, it's quite likely enough to warrant the creation of a new Collection.
3. It's just as likely (or more likely) going to be an accident. As in "we originally put this as CC, but we meant to be open with that _other_ Collection, and not this one". We'll need to handle this case anyway.
4. We do need to keep track of license at time of borrowing so as to not completely screw over the people who used it on that basis...
5. But to keep track of it on a per-BundleVersion level seems like it would introduce a lot of complexity to support a rare operation.

As an example of complexity, let's take discovery. If we do licensing all at the Collection level, then for search purposes something is either public because the Collection is permissively licensed, or it's closed off from browsing unless you belong to that institution. If we do it at the BundleVersion layer, do we make just those specific versions discoverable in search? It really might as well not exist for borrowing with that permissive license if nobody can find it anyway... but then do we work that into the UI so that we're now even seeing subsets of the appropriately licensed versions for an individual Bundle? What does the UI look like for that? What does it look like if the license has changed a couple of times?

It just seems like potentially a lot of edge cases for relatively little gain. Though maybe I'm just not modeling it right in my head...?

> I wasn't part of the early discussions but my feeling is that the goal of tracking license info as a first class thing is to improve discovery of items for re-use (and to enforce restrictions) e.g. "this bundle can be used without modification", "this bundle requires attribution for re-use". I think it would be better if we first came up with a list of scenarios/use-cases that would need to be handled and then figure out what data model would work best and what support needs to be added to the permissions system. Makes sense?

Agreed.

symbolist replied:
> I don't think it's worth the cost to try to track every JavaScript library file in a Bundle and map out their licenses. Blockstore can support a handful of pre-baked licenses at the Collection level -- probably "all rights reserved, only my university gets to use this", and the various flavors of Creative Commons. If a user is going to put a separately licensed third-party piece of content into their Bundle, it should be their responsibility to make sure the licensing is compatible. It should also be their responsibility to include any information needed for proper attribution of authorship.

Makes sense. So basically it must be possible for all content added to a bundle or collection to be offered under the license that is attached to the bundle or collection (depending on which level the license is being attached). 

> As an example of complexity, let's take discovery. If we do licensing all at the Collection level, then for search purposes something is either public because the Collection is permissively licensed, or it's closed off from browsing unless you belong to that institution. If we do it at the BundleVersion layer, do we make just those specific versions discoverable in search? It really might as well not exist for borrowing with that permissive license if nobody can find it anyway... but then do we work that into the UI so that we're now even seeing subsets of the appropriately licensed versions for an individual Bundle? What does the UI look like for that? What does it look like if the license has changed a couple of times?

I am wondering if it would make more sense for authors to specify the policies first and then they are given the option to specify a license label to display that matches those policies. The bundle versions can have the license label attached to them (at the time of creation). However, what we associate with the links are going to be policy objects. This way there can be restrictions on some kinds of policy transitions and it would be simpler to deal with the others.

I am not sure it would work legally to move version 10 of a bundle to a different license (which is not a superset of the earlier license) and then have that license label be displayed when an earlier version is viewed. However, it should be perfectly okay for authors to change (including restrict) the policies around discovery and reuse of all the versions of a bundle or collection in the Blockstore (you are of course free to export or copy an earlier version if you have access to it via a link created before and the license allows it; there is no reliable DRM anyway). I do agree that overall it would be much simpler if these policies applied to all the versions.

I have also created a discovery task for this: https://tasks.opencraft.com/browse/LX-51

ormsbee commented on blockstore/apps/bundles/store.py:
```diff
@@ -0,0 +1,232 @@
+            snapshot_json = json.load(snapshot_file)
+
+        files = {
+            pathname: FileInfo(pathname, file_info[0], file_info[1], codecs.decode(file_info[2], 'hex'))
```
Does `file_info.hash_digest.hex()` not work?

symbolist replied:
Ah okay, I suppose I was doing something wrong then (though I do not remember now). But yeah, that works. Thanks!

ormsbee commented on blockstore/apps/bundles/models.py:
```diff
@@ -0,0 +1,223 @@
+    bundle = models.ForeignKey(
+        Bundle, related_name="versions", related_query_name="version"
+    )
+    version_num = models.PositiveIntegerField()
```
It should be covered in the `unique_together` constraint, though I should move that up to be higher in the class definition.

symbolist replied:
Got it.

ormsbee commented on blockstore/apps/bundles/store.py:
```diff
@@ -0,0 +1,232 @@
+from django.core.files.base import ContentFile, File
+from django.dispatch import Signal
+
+import attr
```
Ack, sorry.

We really need to set up Travis for this...

symbolist replied:
Yup! Going to take care of that soon.

symbolist commented:
Wohoo! 🎉
